paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc

```diff
@@ -0,0 +1,518 @@
+using framework::Tensor;
+using mkldnn::inner_product_forward;
+using mkldnn::memory;
+using mkldnn::prop_kind;
+using mkldnn::stream;
+using platform::MKLDNNDeviceContext;
+using platform::to_void_cast;
+
+template <typename T, typename K>
+class MulPrimitiveFactory {
+ public:
+  explicit MulPrimitiveFactory(const mkldnn::engine &engine)
+      : engine_(engine) {}
+
+  virtual ~MulPrimitiveFactory() {}
+
+  virtual inner_product_forward CreateMulPrimitive(
+      const Tensor *input_x, const Tensor *input_y, Tensor *output,
+      const ExecutionContext &ctx) {
+    /* check format and reorder if need */
+    int x_num_col_dims = ctx.Attr<int>("x_num_col_dims");
+    int y_num_col_dims = ctx.Attr<int>("y_num_col_dims");
+
+    memory::data_type x_dt = framework::ToMKLDNNDataType(input_x->type());
+    memory::data_type y_dt = framework::ToMKLDNNDataType(input_y->type());
+
+    auto x_matrix = UpdateDataFormat(input_x, false,
+                                     /*is_y*/ x_num_col_dims, x_dt, ctx);
+    auto y_matrix = UpdateDataFormat(input_y, true,
+                                     /*is_y*/ y_num_col_dims, y_dt, ctx);
+
+    auto output_dim = output->dims();
+    if (output_dim.size() != 2) {
+      output->Resize({x_matrix.dims()[0], y_matrix.dims()[1]});
+    }
+
+    if (mul_) {
+      UpdateDataPointers(ctx, output, &x_matrix);
+      return *mul_;
+    }
+
+    auto src_desc = CreateMemDescriptor(&x_matrix, x_dt, memory::format::nc);
+    x_input_ = CreateMemory(src_desc, &x_matrix, false /*is_y*/);
+    y_input_ = TransposeInputY(&y_matrix);
+
+    /* for non-quantize mul, output type supposed to be fp32 */
+    auto dst_desc = CreateMemDescriptor<float>(output, memory::format::any);
+
+    mul_ = CreateMulPrimitive(*x_input_, *y_input_, dst_desc, output, ctx);
+    return *mul_;
+  }
+
+ protected:
+  void UpdateDataPointers(const ExecutionContext &ctx, Tensor *out,
+                          const Tensor *in) {
+    x_input_->set_data_handle(const_cast<T *>(in->data<T>()));
+    output_->set_data_handle(out->mutable_data<float>(ctx.GetPlace()));
+    if (out->format() == memory::format::format_undef) {
+      auto output_format = output_->get_primitive_desc().desc().data.format;
+      out->set_format((memory::format)output_format);
+    }
+  }
+
+  Tensor UpdateDataFormat(const Tensor *data, bool is_y, int num_col_dims,
+                          memory::data_type type, const ExecutionContext &ctx) {
+    Tensor x_tmp;
+    Tensor data_matrix;
+
+    memory::format src_fmt = data->format();
+    memory::format dst_fmt;
+    auto dims = framework::vectorize2int(data->dims());
+    auto src_mdesc = CreateMemDescriptor(dims, type, src_fmt);
+
+    if ((data->dims().size() == 4 &&
+         src_fmt != (dst_fmt = memory::format::nchw)) ||
+        (data->dims().size() == 5 &&
+         dst_fmt != (dst_fmt = memory::format::ncdhw))) {
+      auto dst_mdesc = CreateMemDescriptor(dims, type, dst_fmt);
+
+      if (is_y) {
```

Comment:
The only way these two branches differ is the choice of type parameter. This can be solved by templatizing the `UpdateDataFormat` function so that it accepts the data type and uses it for all the `to_void_cast` or `data->data` functions.

```
template <typename DT>
Tensor UpdateDataFormat(const Tensor *data, int num_col_dims,
                          memory::data_type type, const ExecutionContext &ctx)
```
This makes the code shorter and more maintainable.